src/mechanics/orientation.cc

```diff
@@ -0,0 +1,90 @@
+#include <iostream>
+#include <memory>
+
+#include <mechanics/orientation.h>
+
+Model OrientationController::model = Model();
+
+void OrientationController::set_model(Model model1) {
+  model = model1;
+}
+
+OrientationController::OrientationController(Car &c) : car(c) {
+  target = eye<3>();
+  eps_phi = 0.10f;
+
+  finished = false;
+  controls = Input();
+}
+
+void OrientationController::step(float dt) {
+  vec3 omega = dot(transpose(car.o), car.w);
+  mat3 theta = dot(transpose(car.o), target);
+
+  finished = angle_between(car.o, target) < eps_phi;
+
+  vec3 output = model.get_controls(theta, omega);
+
+  controls.roll = output[0];
+  controls.pitch = output[1];
+  controls.yaw = output[2];
+}
+
+Car OrientationController::simulate() {
+
+  Car car_copy = Car(car);
+  auto copy = OrientationController(car_copy);
+  copy.target = target;
+  copy.eps_phi = eps_phi;
+
+  float dt = 0.01666f;
+  for (int t = 0; t < 120; t++) {
+
+    // get the new controls
+    copy.step(dt);
+
+    // and simulate their effect on the car
+    car_copy.step(copy.controls, dt);
+
+    if (copy.finished) break;
+  }
+
+  return car_copy;
+}
+
+Model::Model() = default;
+
+Model::Model(std::vector<float> binary) {
+  weight1 = *reinterpret_cast<mat<20,12>*>(&binary[0]);
+  bias1 = *reinterpret_cast<vec<20>*>(&binary[240]);
+  weight2 = *reinterpret_cast<mat<3,20>*>(&binary[260]);

```

Comment:
Since your binary file is so small, maybe consider just inlining the data

```cpp
weight1 = vec<20>{
1.3253f,
3.2930f,
...
};
```

Reply:
i think inline data makes it harder to read the source.
it's also easier for me to export models to a binary file.